src/coreclr/jit/lclmorph.cpp

```diff
@@ -2426,3 +2426,562 @@ void Compiler::fgExposeLocalsInBitVec(BitVec_ValArg_T bitVec)
+{
+    if (hasConvertFromMask)
+    {
+        // Count the cost of the existing convert mask to vector.
+        weight_t incVal = blockWeight * costOfConvertMaskToVector;
+        JITDUMP("Incrementing currentCost by %f. ", incVal);
+        currentCost += incVal;
+    }
+    else
+    {
+        // Switching would require adding a convert vector to mask.
+        weight_t incVal = blockWeight * costOfConvertVectorToMask;
+        JITDUMP("Incrementing switchCost by %f. ", incVal);
+        switchCost += incVal;
+    }
+    DumpTotalWeight();
+}
+
+//-----------------------------------------------------------------------------
+// UpdateVarWeight: Updates the weighting to take account of a local variable use.
+//
+// Arguments:
+//     hasConvertFromMask - Is this variable converted to a mask when used
+//     blockWeight - Weight of the block the use is contained in
+//
+void Compiler::LCLMasksWeight::UpdateVarWeight(bool hasConvertToMask, weight_t blockWeight)
+{
+    if (hasConvertToMask)
+    {
+        // Count the cost of the existing convert vector to mask.
+        weight_t incVal = blockWeight * costOfConvertVectorToMask;
+        JITDUMP("Incrementing currentCost by %f. ", incVal);
+        currentCost += incVal;
+    }
+    else
+    {
+        // Switching would require adding a convert mask to vector.
+        weight_t incVal = blockWeight * costOfConvertMaskToVector;
+        JITDUMP("Incrementing switchCost by %f. ", incVal);
+        switchCost += incVal;
+    }
+    DumpTotalWeight();
+}
+
+//-----------------------------------------------------------------------------
+// CacheSimdTypes: Cache the simd types of a hwintrinsic
+//
+// Arguments:
+//     op - The HW intrinsic to cache
+//
+void Compiler::LCLMasksWeight::CacheSimdTypes(GenTreeHWIntrinsic* op)
+{
+    CorInfoType newSimdBaseJitType = op->GetSimdBaseJitType();
+    unsigned    newSimdSize        = op->GetSimdSize();
+
+    assert((newSimdBaseJitType != CORINFO_TYPE_UNDEF));
+    assert((simdBaseJitType == CORINFO_TYPE_UNDEF) ||
+           ((simdBaseJitType == newSimdBaseJitType) && (simdSize == newSimdSize)));
+
+    simdBaseJitType = newSimdBaseJitType;
+    simdSize        = newSimdSize;
+}
+
+//-----------------------------------------------------------------------------
+// LCLMasksCheckLCLVarVisitor: Find the user of a lcl var and check if it is a convert to mask
+//
+class LCLMasksCheckLCLVarVisitor final : public GenTreeVisitor<LCLMasksCheckLCLVarVisitor>
+{
+public:
+    enum
+    {
+        DoPostOrder       = true,
+        UseExecutionOrder = true
+    };
+
+    LCLMasksCheckLCLVarVisitor(Compiler* compiler, unsigned lclNum)
+        : GenTreeVisitor<LCLMasksCheckLCLVarVisitor>(compiler)
+        , foundConversion(false)
+        , lclNum(lclNum)
+    {
+    }
+
+    Compiler::fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTree* const lclOp = *use;
```

Comment:
can you add a tiny comment on tree shape that it is trying to look? just for reference?